### PR TITLE
Release of Mattermost, Nextcloud and Dashboard

### PR DESCRIPTION
**Pull request by zonradkuse**
This PR contains all commits necessary for a production deployment of mattermost and nextcloud.
Until then we still need
- [x] autossh for ldap
- [x] automatic update method for mattermost
- [x] dashboard
- [x] nextcloud
- [x] backups
- [x] netdata
- [x] security:
  - [x] fail2ban
  - [x] rkhunter
  - [x] no password logins

---

After production deployment:

- [ ] nextcloud
  - [ ] stop nextcloud.service
  - [ ] mysql migration
  - [ ] `MariaDB [nextcloud]> update oc_appconfig set configvalue = '172.16.238.1' where appid = 'user_ldap' and configkey = 'ldap_host';`
  - [ ] `docker exec -ti --user www-data nextcloud_nextcloud_1 /bin/bash` 
    - [ ] `php occ upgrade`
  - [ ] files migration
  - [ ] enable calendar, contacts, deck
  - [ ] install LDAP Contacts
  - [ ] install [files_readmemd](https://apps.nextcloud.com/apps/files_readmemd)

- [ ] mattermost
  - [ ] mysql migration
  - [ ] file migration
  - no ldap migration necessary - oauth stays on old server for now

### CODE REVIEW

**Comment by marcelklehr** [comment]
I tired to ssh to webtropia in the past, but wasn't able to for some reason, even though the ssh key was authorized. :/

**Comment by marcelklehr** [comment]
>  automatic update method for mattermost

What do you mean by 'automatic'?

**Comment by zonradkuse** [comment]
> What do you mean by 'automatic'?

given an old version of mattermost we do a data preserving update including backups before initiating the update procedure. Basically, what is described in https://docs.mattermost.com/administration/upgrade.html

File: site.yml [hunk 1]
```diff
@@ -22,16 +21,17 @@
...
+    - geerlingguy.nodejs
+    - firewalld
+    #- php
+    #- geerlingguy.composer
     - acme-nginx
-    - tomcat
+    #- tomcat
     - nextcloud
     - mattermost
     - dashboard
```
Comment (marcelklehr) [inline site.yml]: Do you also want to deploy dashboard immediately?

File: deploy.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+	ANSIBLE_CALL="${ANSIBLE_BASE_CALL} ${PRODUCTION_INVENTORY_PATH}"
+fi
+
+if [[ $BRANCH = "staging" ]]; then
+	ANSIBLE_CALL="${ANSIBLE_BASE_CALL} ${STAGING_INVENTORY_PATH}"
+fi
+
+echo ${ANSIBLE_CALL}
+${ANSIBLE_CALL}
```
Comment (marcelklehr) [inline deploy.sh]: :+1:

(The Warning could use a little more fun. Something like this: https://www.youtube.com/watch?v=5p8wTOr8AbU)

File: site.yml (hunk 1 above)
Comment (zonradkuse) [reply site.yml]: that was actually by accident. We could actually do this, though. It just requires the ldap connection.

Comment (marcelklehr) [reply site.yml]: Yeah, and it's quite stable, so why not.

Comment (marcelklehr) [reply site.yml]: dashboard needs php and remi-repo, though

Comment (zonradkuse) [reply site.yml]: done :)

File: roles/ldap-ssh/templates/autossh-ldap.service.j2 [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Nextcloud Apache server
```
[the remark this answers is not in the document]
Comment (zonradkuse) [reply roles/ldap-ssh/templates/autossh-ldap.service.j2]: https://www.everythingcli.org/ssh-tunnelling-for-fun-and-profit-autossh/#gfm-8

[the remark this answers is not in the document]
Comment (zonradkuse) [reply roles/ldap-ssh/templates/autossh-ldap.service.j2]: #82a193c cool! :)

File: roles/php/tasks/main.yml [hunk 4]
```diff
@@ -13,4 +13,4 @@
     line: "group = nginx"
 
 - import_role:
-    name: - geerlingguy.composer
```
Comment (zonradkuse) [inline roles/php/tasks/main.yml]: Next time I will do the syntax check, I promise :D

File: roles/ldap-ssh/templates/autossh-ldap.service.j2 [hunk 5]
```diff
@@ -3,7 +3,6 @@ Description=AutoSSH tunnel for LDAP
 After=network.target
 
 [Service]
-Environment="AUTOSSH_GATETIME=0"
```
Comment (zonradkuse) [inline roles/ldap-ssh/templates/autossh-ldap.service.j2]: We need to double check that removing this survives a reboot

Comment (marcelklehr) [reply roles/ldap-ssh/templates/autossh-ldap.service.j2]: Mhyea. IMO it's a shitty flag and enabling it to survive boot-time is kind of a hack. Maybe we can simply use a different `After=` target?

Comment (zonradkuse) [reply roles/ldap-ssh/templates/autossh-ldap.service.j2]: hm. mattermost/nextcloud/...?

**Comment by marcelklehr** [comment]
> provisioning of valid ssh-keys per server in order to ssh to legacy server

This seems like overkill for a temporary solution, I think.

**Comment by zonradkuse** [comment]
> > provisioning of valid ssh-keys per server in order to ssh to legacy server
>
> This seems like overkill for a temporary solution, I think.

It is. I wanted to do this already for the backup but found that it is rather an overkill. Instead I abort if I don't find a ssh key and assume that it is setup for password-less login.

**Comment by zonradkuse** [comment]
are we ready for the release? I would remove the WIP tag then so we can merge to master after a review.

**Comment by marcelklehr** [comment]
:rocket:

**Comment by zonradkuse** [comment]
Okay, we are ready to go :)

**Pull request merged**